File: coding/server/engine/channelRouter.js
/**
 * Channel Router & Adaptive Capacity Module
 * Maps farmer digital profile (device type, network quality, tech literacy)
 * to appropriate contact channel and default UI experience mode.
 */

/**
 * Recommends the optimal delivery channel for advisories and alerts
 * @param {Object} farmer 
 * @returns {"in_app_voice_and_text" | "ivr_or_sms"}
 */
function getRecommendedChannel(farmer) {
  if (!farmer) return "ivr_or_sms";

  // If farmer has a smartphone and at least moderate connectivity, deliver rich in-app voice/text
  if (farmer.device_type === "smartphone" && farmer.network_quality !== "poor") {
    return "in_app_voice_and_text";
  }

  // If farmer relies on feature phone / basic smartphone OR has poor rural connectivity, deliver via voice IVR / SMS
  if (
    farmer.device_type === "feature_phone" ||
    farmer.device_type === "basic_smartphone" ||
    farmer.network_quality === "poor"
  ) {
    return "ivr_or_sms";
  }

  return "ivr_or_sms";
}

/**
 * Determines whether the UI should pre-select "Assisted Mode" (Kisan Mitra / CSC)
 * or "Self-Service Mode"
 * @param {Object} farmer 
 * @returns {"assisted" | "self"}
 */
function getDefaultUIMode(farmer) {
  if (!farmer) return "assisted";
  if (farmer.tech_literacy === "low" || farmer.device_type === "feature_phone") {
    return "assisted";
  }
  return "self";
}

module.exports = {
  getRecommendedChannel,
  getDefaultUIMode
};

File: coding/server/engine/channel_router.py
"""
Channel Router & Adaptive Capacity Module (Python)
Maps farmer digital profile (device type, network quality, tech literacy)
to appropriate contact channel and default UI experience mode.
"""

def get_recommended_channel(farmer: dict) -> str:
    """
    Recommends the optimal delivery channel for advisories and alerts:
    - 'in_app_voice_and_text' for smartphone with moderate/good connectivity
    - 'ivr_or_sms' for feature phone, basic smartphone, or poor connectivity
    """
    if not farmer:
        return "ivr_or_sms"
    
    device_type = farmer.get("device_type", "feature_phone")
    network_quality = farmer.get("network_quality", "poor")
    
    if device_type == "smartphone" and network_quality != "poor":
        return "in_app_voice_and_text"
        
    if device_type in ["feature_phone", "basic_smartphone"] or network_quality == "poor":
        return "ivr_or_sms"
        
    return "ivr_or_sms"


def get_default_ui_mode(farmer: dict) -> str:
    """
    Determines whether the UI should pre-select "Assisted Mode" (Kisan Mitra / CSC)
    or "Self-Service Mode"
    """
    if not farmer:
        return "assisted"
        
    tech_literacy = farmer.get("tech_literacy", "low")
    device_type = farmer.get("device_type", "feature_phone")
    
    if tech_literacy == "low" or device_type == "feature_phone":
        return "assisted"
        
    return "self"
